backends/lean/Aeneas/Progress/ProgressStar.lean

```diff
@@ -0,0 +1,326 @@
+    -- Since (· >>= pure) = id, we treat a result as a bind on id
+    onBind expr resultName (pure {})
+
+  onBind _curr name processRest := do
+    trace[ProgressStar] s!"onBind: encountered {←ppExpr _curr}"
+    if let some {usedTheorem, ..} ← tryProgress then
+      trace[ProgressStar] s!"onBind: Can make progress! Binding {name}"
+      let (preconditionTacs, unsolved) ← trySolvePreconditions
+      if ¬ preconditionTacs.isEmpty then
+        trace[ProgressStar] s!"onBind: Found {preconditionTacs.size} preconditions, left {unsolved.size} unsolved"
+      let ids ← getIdsFromUsedTheorem name usedTheorem
+      if ¬ ids.isEmpty && ¬ (←getGoals).isEmpty then
+        replaceMainGoal [← renameInaccessibles (← getMainGoal) ids] -- NOTE: Taken from renameI tactic
+      let currTac ← if ids.isEmpty 
+        then `(tactic| progress with $(←usedTheorem.toSyntax))
+        else `(tactic| progress with $(←usedTheorem.toSyntax) as ⟨$ids,*⟩)
+      let restInfo ← processRest
+      return {
+        script := #[currTac]++ preconditionTacs, -- TODO: Optimize
+        unsolvedPreconditions := unsolved.toList
+      } ++ restInfo
+    else return {script:=#[]}
+
+  onBif bfInfo toBeProcessed := do
+    trace[ProgressStar] s!"onBif: encountered {bfInfo.kind}"
+    if (←getGoals).isEmpty then 
+      trace[ProgressStar] s!"onBif: no goals to be solved!"
+      -- Tactic.focus fails if there are no goals to be solved.
+      return {}
+    Tactic.focus do
+      let splitStx ← `(tactic| split)
+      evalSplit splitStx
+      -- 
+      let subgoals ← getUnsolvedGoals
+      trace[ProgressStar] s!"onBif: Bifurcation generated {subgoals.length} subgoals"
+      unless subgoals.length == toBeProcessed.size do
+        throwError s!"onBif: Expected {toBeProcessed.size} cases, found {subgoals.length}"
+      -- Gather suggestions from branches
+      let mut unsolvedGoals: List (MVarId) := []
+      let mut info := {script:=#[splitStx]}
+      for (sg, (names, processBr)) in subgoals.zip toBeProcessed.toList do
+        setGoals [sg]
+        let branchInfo ← processBr
+
+        let branchTacs ← if branchInfo.script.isEmpty 
+          then pure #[←`(tactic| skip)]
+          else pure branchInfo.script
+        let caseArgs := makeCaseArgs (←sg.getTag) names
+        info := {info ++ branchInfo with 
+          script:=info.script.push <| ←`(tactic| case' $caseArgs => $branchTacs*),
+        }
+        unsolvedGoals := unsolvedGoals ++ (←getUnsolvedGoals)
+
+      setGoals unsolvedGoals
+      return info
+
+  tryProgress := do
+    try some <$> Progress.evalProgress none none #[]
+    catch _ => pure none
+
+  trySolvePreconditions: TacticM (Array Syntax.Tactic × Array MVarId) := do
+    -- NOTE: Do I make the assumption that the preconditions appear before the final lemma, or
+    -- that the names of the preconditions contain the final lemma as a prefix.
+    -- For now, we have ←getUnsolvedGoals = preconditions ++ [final]
```

Comment:
I think we should update `evalProgress` so that it has type: `Array MVarId × Option MVarId` (list of preconditions + remaining goal, if there is still one)